File: leetcode/rust-leetcode/src/238.product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![1; nums.len()];
        for i in 1..nums.len() {
            result[i] = result[i - 1] * nums[i - 1];
        }
        let mut right = 1;
        for i in (0..nums.len()).rev() {
            result[i] *= right;
            right *= nums[i];
        }
        result
    }
}

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut products = vec![1; n];
        for i in 1..n {
            products[i] = nums[i-1] * products[i-1];
        }
        let mut right_prod = 1;
        for i in (0..n).rev() {
            products[i] *= right_prod;
            right_prod *= nums[i];
        }
        products
    }
}

File: leetcode/python/238.product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        products = [1 for _ in nums]
        product = 1
        for i, n in enumerate(nums):
            products[i] = product
            product = product * n
        product = 1
        N = len(nums)
        for i, n in enumerate(nums[::-1]):
            products[N-i-1] = products[N-i-1] * product
            product = product * n
        return products

